src/metabase/upload.clj

```diff
@@ -612,7 +620,12 @@
           normed-name->field (m/index-by (comp normalize-column-name :name)
                                          (t2/select :model/Field :table_id (:id table) :active true))
           normed-header      (map normalize-column-name header)
-          create-auto-pk?    (not (contains? normed-name->field auto-pk-column-name))
+          ;; Create auto-pk columns for drivers that supported uploads before auto-pk columns
+          ;; were introduced by metabase#36249. Otherwise we can assume that the table was created
+          ;; with an auto-pk column, or it has been deliberately removed.
+          create-auto-pk?    (and
```

Comment:
This is a very high level part of the spec, so I don't think it's appropriate for it to be squirreled away inside a large mutating method. I think extracting a top level function like `(append-pk-on-upload? database-egine)` is a good idea.

Reply:
We tend to write pretty thorough docstrings, so I guess putting it there is a good idea. My main concern is that it's easy to lose track of every detail in large comments, and they can fall out of sync with what the code really does.

I'm not sure I see how the problem is the same - if we define a top level function then it shows up in the code outline as a first class concept, so more discoverable and atomic. On the tin instead of inside it. The same principle as extracting constants versus commenting on magic variables, even if they're only referenced once.

If you separated the business logic and implementation details into separate namespaces it would stand out even more prominently. In the general case it's also more testable, or usable by tests to determine what the expected side effects here.

If docstrings are king then it's also easier to keep that function's docstring uptodate as it'd be right next to the map literal.

I'm not going to push hard on this now, but it's got me thinking about how we structure things. I'm sure my thinking will evolve as I get more familiar with the codebase.

Reply:
I'm genuinely fascinated about how differently we think about code structure and documentation, look forward to discussing it more as we write more code together. Happy to take guidance on our existing conventions as a start.

Moving the logic to the multimethod certainly ticks my boxes on pushing the business logic out of the implementation details, but I have two passing comments.

1. I thought you gave a really good reason not to make this extensible at first - only this subset of drivers will have created non-pk tables in the past, so it's a moot point for the others. It seems unfortunate to bloat the public interface in this case, even if in practice new drivers will all use the default implementation.

2. I'm not that comfortable with having `csv` in the method name. This aspect of the behavior has nothing to do with the source file format, and we may support addition ones in future, e.g. parquet.

Reply:
On 1, I'm a little torn but I think the more "correct" thing to do is follow Ngoc's suggestion to add the multimethod, since it's possible (though unlikely) 3rd party drivers could have implemented uploads in 47. It's a little unfortunate the public interface has increased but it's a small cost to pay as you said, new drivers will use the default implementation. Since I was extracting the function anyway, it seemed worth doing right.

On 2 I agree. I'll make the change.